tcp msg ring panic

...attempting to close nonexistent net.Conn obj.

```
2015/08/28 05:12:11 msgToNode error: write tcp 10.X.X.X:8001: connection reset by peer
2015/08/28 05:12:11 conn struct: &ring.ringConn{state:2, addr:"10.X.X.X:8001", conn:(*net.TCPConn)(0xc465116380), reader:(*ring.timeoutReader)(0xc448bc2fe0), writerLock:sync.Mutex{state:1, sema:0x0},
 writer:(*ring.timeoutWriter)(0xc448bc30a0)}
015/08/28 05:12:11 handleForever error: read tcp 10.X.X.X:8001: use of closed network connection
2015/08/28 05:12:11 conn struct: &ring.ringConn{state:1, addr:"10.X.X.X:8001", conn:net.Conn(nil), reader:(*ring.timeoutReader)(nil), writerLock:sync.Mutex{state:0, sema:0x0}, writer:(*ring.timeoutWr
iter)(nil)}
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x20 pc=0x51bd43]

goroutine 1485 [running]:
github.com/gholt/ring.(*TCPMsgRing).disconnection(0xc20807a180, 0xc208079830, 0x10)
        /root/go/src/github.com/gholt/ring/tcp_msg_ring.go:138 +0x1d3
```

That error's actually on line 137 at https://github.com/gholt/ring/blob/master/tcp_msg_ring.go#L137 but I added some debug info locally.

--- a/tcp_msg_ring.go
+++ b/tcp_msg_ring.go
@@ -132,6 +132,7 @@ func (m *TCPMsgRing) disconnection(addr string) {
 	m.lock.Lock()
 	conn := m.conns[addr]
 	delete(m.conns, addr)
+	m.lock.Unlock()
 	if conn != nil {
 		if conn.conn != nil {
 			err := conn.conn.Close()
@@ -141,7 +142,6 @@ func (m *TCPMsgRing) disconnection(addr string) {
 		}
 
 	}
-	m.lock.Unlock()
 }
 
 func (m *TCPMsgRing) handshake(conn *ringConn) error {

--- a/tcp_msg_ring.go
+++ b/tcp_msg_ring.go
@@ -132,10 +132,16 @@ func (m *TCPMsgRing) disconnection(addr string) {
 	m.lock.Lock()
 	conn := m.conns[addr]
 	delete(m.conns, addr)
-	m.lock.Unlock()
 	if conn != nil {
-		conn.conn.Close()
+		if conn.conn != nil {
+			err := conn.conn.Close()
+			if err != nil {
+				log.Println("tcp msg ring disconnection close err:", err)
+			}
+		}
+
 	}
+	m.lock.Unlock()
 }
 
 func (m *TCPMsgRing) handshake(conn *ringConn) error {

